src/contracts/types/contract.ts

```diff
@@ -0,0 +1,38 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
+/** A generic contract input. */
+export type ContractInput = {function: string; value: any};
+
+/** A contract action, that is a caller and their input. */
+export type ContractAction<Input extends ContractInput = ContractInput> = {
+  input: Input;
+  caller: string;
+};
+
+/** A contract state. */
+export type ContractState<Mode extends {whitelists: string[]; circuits: string[]} = {whitelists: []; circuits: []}> = {
+  canEvolve: boolean;
+  evolve?: string;
+  owner: string;
+  // proofs mode
+  isProofRequired: {[name in Mode['circuits'] extends [] ? string : Mode['circuits'][number]]: boolean};
```

Comment:
this magicaly line simply lets the dictionary to have arbitrary keys if `Mode['circuits']` is `[]`, and if it is another array it will only allow dictionary keys that are in that array.